src/transformers/SingleLoopTransformation.cc

```diff
@@ -0,0 +1,286 @@
+    for (PTRef auxVar : edgeVariables.auxiliaryVars) {
+        this->auxiliaryVariablesSeen.push(auxVar);
+    }
+
+    // TODO: prepare the substitution map in advance!
+    auto const & stateVars = edgeVariables.stateVars;
+    for (unsigned int i = 0; i < stateVars.size(); ++i) {
+        VarPosition varPosition {source, i};
+        auto it = positionVarMap.find(varPosition);
+        assert(it != positionVarMap.end());
+        substitutionsMap.insert({stateVars[i], it->second});
+    }
+
+    auto const & nextStateVars = edgeVariables.nextStateVars;
+    for (unsigned int i = 0; i < nextStateVars.size(); ++i) {
+        VarPosition varPosition {target, i};
+        auto it = positionVarMap.find(varPosition);
+        assert(it != positionVarMap.end());
+        substitutionsMap.insert({nextStateVars[i], timeMachine.sendVarThroughTime(it->second, 1)});
+    }
+
+    // Translate the constraint
+    PTRef translatedConstraint = TermUtils(logic).varSubstitute(edge.fla.fla, substitutionsMap);
+    PTRef sourceLocationVar = source == graph.getEntry() ? logic.getTerm_true() : locationVarMap.at(source);
+    PTRef targetLocationVar = locationVarMap.at(target);
+    PTRef updatedLocation = [&]() {
+        vec<PTRef> args;
+        args.capacity(locationVarMap.size() * 2);
+        for (auto && entry : locationVarMap) {
+            if (entry.second != targetLocationVar) {
+                args.push(logic.mkNot(timeMachine.sendVarThroughTime(entry.second, 1)));
+            }
+            if (entry.second != sourceLocationVar) {
+                args.push(logic.mkNot(entry.second));
+            }
+        }
+        return logic.mkAnd(std::move(args));
+    }();
+    PTRef frameEqualities = [&]() {
+        vec<PTRef> equalities;
+        for (auto && entry : positionVarMap) {
+            if (entry.first.vertex == target) { continue; }
+            PTRef var = timeMachine.sendVarThroughTime(entry.second, 1);
+            equalities.push(logic.mkEq(var, logic.getDefaultValuePTRef(var)));
+        }
+        return logic.mkAnd(std::move(equalities));
+    }();
+    vec<PTRef> components {
+        sourceLocationVar,
+        translatedConstraint,
+        timeMachine.sendVarThroughTime(targetLocationVar, 1),
+        updatedLocation,
+        frameEqualities
+    };
+    return logic.mkAnd(std::move(components));
+}
+
+}
+
+SingleLoopTransformation::TransformationResult SingleLoopTransformation::transform(const ChcDirectedGraph & graph) {
+    Logic & logic = graph.getLogic();
+    TimeMachine timeMachine(logic);
+    auto vertices = graph.getVertices();
+    // MB: It is useful to have exit location, so we do not remove exit from the vertices
+    vertices.erase(std::remove_if(vertices.begin(), vertices.end(), [&](auto vertex) {
+                       return vertex == graph.getEntry();
+                   }), vertices.end());
+    LocationVarMap locationVars;
+    locationVars.reserve(vertices.size());
+    for (auto vertex : vertices) {
```

Comment:
There is no location variable for entry point.
The exit node has a location variable because that becomes the new safety property, that this variable never becomes true. Maybe this translation could be optimized, but that is for another time.